twine/commands/upload.py

```diff
@@ -22,30 +22,26 @@
 
 
 def skip_upload(response, skip_existing, package):
-    filename = package.basefilename
-    msg_400 = (
-        # Old PyPI message:
-        f'A file named "{filename}" already exists for',
-        # Warehouse message:
-        'File already exists',
-        # Nexus Repository OSS message:
-        'Repository does not allow updating assets',
-    )
-    msg_403 = 'Not enough permissions to overwrite artifact'
+    if not skip_existing:
+        return False
+
+    status = response.status_code
+    reason = getattr(response, 'reason', '').lower()
+    text = getattr(response, 'text', '').lower()
+
     # NOTE(sigmavirus24): PyPI presently returns a 400 status code with the
     # error message in the reason attribute. Other implementations return a
-    # 409 or 403 status code. We only want to skip an upload if:
-    # 1. The user has told us to skip existing packages (skip_existing is
-    #    True) AND
-    # 2. a) The response status code is 409 OR
-    # 2. b) The response status code is 400 AND it has a reason that matches
-    #       what we expect PyPI or Nexus OSS to return to us. OR
-    # 2. c) The response status code is 403 AND the text matches what we
-    #       expect Artifactory to return to us.
-    return (skip_existing and (response.status_code == 409 or
-            (response.status_code == 400 and
-             response.reason.startswith(msg_400)) or
-            (response.status_code == 403 and msg_403 in response.text)))
+    # 403 or 409 status code.
+    return any([
```

Comment:
So in this case, we generate a list before we can call `any` which can otherwise be lazy. What if, we choose a lazier way to implement this given tat `in` with strings can be slow and people expect their CLIs to be quick. (Especially considering `text` can be a giant HTML page and really really slow to scan through.) In that case, I'd prefer if this looked like:

```py
conditions = [(400, 'already exist', reason), (400, 'updating asset', reason), (403, 'overwrite artifact', text)]
return status == 409 or any(status == status_code and search_for in string for (status_code, search_for, string) in conditions)
```

`any` will consume the generator and then we won't unnecessarily scan through a potentially giant document _every time_ we see a 403

Reply:
Thanks, @sigmavirus24. I'm usually all about laziness; just missed it here. I took a different approach in https://github.com/pypa/twine/pull/510/commits/e366401f322e3f314dbcf4cf781a9cd3aa8caba5 (that's actually similar to the prior implementation).